jerry-core/ecma/operations/ecma-async-generator-object.c

```diff
@@ -169,10 +329,129 @@ ecma_async_generator_finalize (vm_executable_object_t *async_generator_object_p,
+      {
+        ecma_free_value (value);
+        return result;
+      }
+
+      bool done = ecma_op_to_boolean (result);
+      ecma_free_value (result);
+      result = ecma_op_object_get_by_magic_id (result_obj_p, LIT_MAGIC_STRING_VALUE);
+      ecma_free_value (value);
+
+      if (ECMA_IS_VALUE_ERROR (result))
+      {
+        return result;
+      }
+
+      if (!done)
+      {
+        ECMA_ASYNC_YIELD_ITERATOR_SET_STATE (async_generator_object_p, NEXT_VALUE);
+        return ecma_promise_async_await ((ecma_extended_object_t *) async_generator_object_p, result);
+      }
+
+      if (ECMA_ASYNC_YIELD_ITERATOR_GET_STATE (async_generator_object_p) == ECMA_ASYNC_YIELD_ITERATOR_AWAIT_NEXT)
+      {
+        ECMA_ASYNC_YIELD_ITERATOR_END (async_generator_object_p);
+        return result;
+      }
+
+      ECMA_ASYNC_YIELD_ITERATOR_END (async_generator_object_p);
+      async_generator_object_p->frame_ctx.byte_code_p = opfunc_resume_executable_object_with_return;
+      return result;
+    }
+    case ECMA_ASYNC_YIELD_ITERATOR_AWAIT_RETURN:
+    {
+      ecma_object_t *obj_p = ecma_get_object_from_value (async_generator_object_p->frame_ctx.block_result);
+      ecma_value_t result = ecma_op_object_get_by_magic_id (obj_p, LIT_MAGIC_STRING_RETURN);
+
+      if (ECMA_IS_VALUE_ERROR (result))
+      {
+        ecma_free_value (value);
+        return result;
+      }
+
+      if (result == ECMA_VALUE_UNDEFINED)
+      {
+        ECMA_ASYNC_YIELD_ITERATOR_END (async_generator_object_p);
+        async_generator_object_p->frame_ctx.byte_code_p = opfunc_resume_executable_object_with_return;
+        return value;
+      }
+
+      result = ecma_async_yield_call (result,
+                                      async_generator_object_p,
+                                      value,
+                                      ECMA_ERR_MSG ("Iterator return() is not callable."));
+      ecma_free_value (value);
+
+      if (ECMA_IS_VALUE_ERROR (result))
+      {
+        return result;
+      }
+
+      JERRY_ASSERT (result == ECMA_VALUE_UNDEFINED);
+      ECMA_ASYNC_YIELD_ITERATOR_CHANGE_STATE (async_generator_object_p, RETURN, NEXT_RETURN);
+      return ECMA_VALUE_UNDEFINED;
+    }
+    case ECMA_ASYNC_YIELD_ITERATOR_AWAIT_NEXT_VALUE:
+    {
+      ECMA_ASYNC_YIELD_ITERATOR_CHANGE_STATE (async_generator_object_p, NEXT_VALUE, OPERATION);
+      opfunc_async_generator_yield ((ecma_extended_object_t *) async_generator_object_p, value);
+      return ECMA_VALUE_UNDEFINED;
+    }
+    case ECMA_ASYNC_YIELD_ITERATOR_AWAIT_OPERATION:
+    {
+      /* Currently this is always a throw exception case. */
+      ecma_value_t result = ecma_async_yield_throw (async_generator_object_p, value);
+      ecma_free_value (value);
+      return result;
+    }
+    default:
+    {
+      JERRY_ASSERT (ECMA_ASYNC_YIELD_ITERATOR_GET_STATE (async_generator_object_p)
+                    == ECMA_ASYNC_YIELD_ITERATOR_AWAIT_CLOSE);
+
+      if (!ecma_is_value_object (value))
```

Comment:
How about saving this into a boolean? Therefore the double free/raise error calls can be reduced into single ones.